torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -662,10 +716,54 @@ class TORCH_CUDA_CU_API IterDomain : public Val {
     return definition() == nullptr;
   }
 
+  bool isInstruction() const {
+    return is_instruction_;
+  }
+
+  bool isWarpMapped() const {
+    return is_warp_mapped_;
+  }
+
  protected:
   friend TensorDomain;
   friend ReplayTransformations;
   friend IndexReferenceReplay;
+  friend mma_util::WarpMmaSwizzler;
+
+ private:
+  //! Used by WarpMmaSwizzler, marks that this id represents a
+  //!  instruction loop, mma use only.
+  //!
+  //! An instruction loop can be considered a generalization of
+  //!  vectorization. It also represents a loop that's implemented
+  //!  by an instruction and should not be realized by codegen and
+  //!  cannot be inlined with.
+  //! As an example, if a mma macro, call it mma_eg implements:
+  //!  for m in M
+  //!    for n in N
+  //!      for k in K
+  //!         C[m,n] += A[m,k]*B[k,n],
+  //! But the generated code should simply be:
+  //!  mma_eg(C,A,B)
+  //! without the 3 level loopnest, i.e. they're instruction loops.
+  //!
+  //! In the actual mma macros, the loopnests it implements is a
+  //!  transformed version of above to match the mma swizzle.
+  //!  So it's different implicit loopnest for different macros.
+  //!  WarpMmaSwizzler will label the instruction loops case-by-case.
+  void toInstruction() {
+    is_instruction_ = true;
+  }
+
+  //! Used by WarpMmaSwizzler, this is an utility for WarpMmaSwizzler
```

Comment:
I'm still not sure what this is for. It seems that `isWarpMapped()` is only used for validation.

Can this be inferred from MMA expressions? Looking at the usage of `toWarpMapped` in mma_utils.cpp, detecting those domains seems trivial.

Reply:
I will rename this to `isMmaSwizzled` if it sounds clearer what it does. 

At current stage this tag is just used as a "locking" and "validation" mechanism. 

2 things need to be checked to make sure mma generates the right data:
  1. The correct axes are being swizzled with mma pattern. 
  2. The mma pattern is correct for the operand/macro etc.

The validation method at this point could just be an intermediate state where we could add more complexity if needed: 
  `WarpMmaSwizzler` will check 1 and then will apply the correct swizzle pattern. Then we "lock" these axes so that they stay correct until lowering.

`isMmaSwizzled` is the "locking" part that only `WarpMmaSwizzler` can initiate. This prevent cases like `a scheduler applies mma swizzle and then merge and bind the innermost ids to BIDx`, which just be invalid.